lib/kaur/result.ex

```diff
@@ -91,6 +91,52 @@ defmodule Kaur.Result do
   def from_value(value), do: ok(value)
 
   @doc """
+  Converts an `Ok` value to an `Error` value if the `predicate` is not valid.
+
+  ## Examples
+
+      iex> res = Kaur.Result.ok(10)
+      ...> Kaur.Result.keep_if(res, &(&1 > 5))
+      {:ok, 10}
+
+      iex> res = Kaur.Result.ok(10)
+      ...> Kaur.Result.keep_if(res, &(&1 > 10), "must be > of 10")
+      {:error, "must be > of 10"}
+
+      iex> res = Kaur.Result.error(:no_value)
+      ...> Kaur.Result.keep_if(res, &(&1 > 10), "must be > of 10")
+      {:error, :no_value}
+  """
+  @spec keep_if(result_tuple, (any -> boolean), any) :: result_tuple
+  def keep_if(result, predicate, error_message \\ :invalid)
+  def keep_if({:error, _} = error, _predicate, _error_message), do: error
+  def keep_if({:ok, value} = ok, predicate, error_message) do
+    if predicate.(value), do: ok, else: error(error_message)
+  end
+
+  @doc """
+  Converts an `Ok` value to an `Error` value if the `predicate` is valid.
+
+  ## Examples
+
+      iex> res = Kaur.Result.ok([])
+      ...> Kaur.Result.keep_unless(res, &Enum.empty?/1)
+      {:error, :invalid}
+
+      iex> res = Kaur.Result.ok([1])
+      ...> Kaur.Result.keep_unless(res, &Enum.empty?/1)
+      {:ok, [1]}
+
+      iex> res = Kaur.Result.ok([])
+      ...> Kaur.Result.keep_unless(res, &Enum.empty?/1, "list cannot be empty")
+      {:error, "list cannot be empty"}
+  """
+  @spec keep_unless(result_tuple, (any -> boolean), any) :: result_tuple
+  def keep_unless(result, predicate, error_message \\ :invalid) do
```

Comment:
Which naming do you prefer? `keep_unless` or `reject_if`?

Reply:
In fact, I prefer `reject_if`, but IMHO, keep_unless is "more" in the if/unless logic ... What's your point ?

Reply:
We have `filter` and `reject` in `Enum` when we want to keep or reject things so I was wondering if we should follow this same naming or not.

I don't have real opinion, just asking what would be the best naming

Reply:
Ok, agreed ;)